server/src/main/java/org/elasticsearch/search/aggregations/AggregationPhase.java

```diff
@@ -29,30 +31,49 @@ public static void preProcess(SearchContext context) {
-            bucketCollector.preCollection();
-        } catch (IOException e) {
-            throw new AggregationInitializationException("Could not initialize aggregators", e);
-        }
-        final Collector collector;
-        if (context.aggregations().factories().context() != null
-            && context.aggregations().factories().context().isInSortOrderExecutionRequired()) {
-            TimeSeriesIndexSearcher searcher = new TimeSeriesIndexSearcher(context.searcher(), getCancellationChecks(context));
-            searcher.setMinimumScore(context.minimumScore());
-            searcher.setProfiler(context);
+        final Supplier<BucketCollector> bucketCollectorSupplier = () -> {
             try {
-                searcher.search(context.rewrittenQuery(), bucketCollector);
+                Aggregator[] aggregators = context.aggregations().factories().createTopLevelAggregators();
+                context.aggregations().aggregators(aggregators);
+                BucketCollector bucketCollector = MultiBucketCollector.wrap(true, List.of(aggregators));
+                bucketCollector.preCollection();
+                return bucketCollector;
             } catch (IOException e) {
-                throw new AggregationExecutionException("Could not perform time series aggregation", e);
+                throw new AggregationInitializationException("Could not initialize aggregators", e);
             }
-            collector = BucketCollector.NO_OP_COLLECTOR;
+        };
+        final Supplier<Collector> collectorSupplier;
+        if (context.aggregations().isInSortOrderExecutionRequired()) {
+            executeInSortOrder(context, bucketCollectorSupplier.get());
+            collectorSupplier = () -> BucketCollector.NO_OP_COLLECTOR;
         } else {
-            collector = bucketCollector.asCollector();
+            collectorSupplier = () -> bucketCollectorSupplier.get().asCollector();
+        }
+        context.aggregations().registerAggsCollectorManager(new CollectorManager<>() {
+            @Override
+            public Collector newCollector() {
+                return collectorSupplier.get();
+            }
+
+            @Override
+            public Void reduce(Collection<Collector> collectors) {
+                // we cannot run post-collection method here because we need to do it after the optional timeout
+                // has been removed from the index searcher. Therefore, we delay this processing to the
+                // AggregationPhase#execute method.
+                return null;
+            }
+        });
+    }
+
+    private static void executeInSortOrder(SearchContext context, BucketCollector collector) {
+        TimeSeriesIndexSearcher searcher = new TimeSeriesIndexSearcher(context.searcher(), getCancellationChecks(context));
+        searcher.setMinimumScore(context.minimumScore());
+        searcher.setProfiler(context);
+        try {
+            searcher.search(context.rewrittenQuery(), collector);
```

Comment:
I assume that we are not planning to move this search call to use a collector manager?

Reply:
no, time series execution cannot work with concurrency as we need access to all segments in order to sort the data.